pyface/data_view/abstract_value_type.py

```diff
@@ -0,0 +1,182 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" Provides an AbstractValueType ABC for Pyface data models.
+
+This module provides an ABC for data view value types, which are responsible
+for adapting raw data values as used by the data model's ``get_value`` and
+``set_value`` methods to the data channels that the data view expects, such
+as text, color, icons, etc.
+
+It is up to the data view to take this standardized data and determine what
+and how to actually display it.
+"""
+
+from traits.api import ABCHasStrictTraits, Event, observe
+
+
+class AbstractValueType(ABCHasStrictTraits):
+    """ A value type converts raw data into data channels.
+
+    The data channels are editor value, text, color, image, and description.
+    The data channels are used by other parts of the code to produce the actual
+    display.
+
+    Subclasses should mark traits that potentially affect the display of values
+    with ``update=True`` metdadata, or alternatively fire the ``updated``
+    event when the state of the value type changes.
+    """
+
```

Comment:
There will (once we have proper support code) be other get/sets for color data, image data, etc. so this will not end up being just text vs. editable values.  (Although I suspect that most of the others will not be settable).  It may be helpful to read up on the QAbstractItemModel's data() method: https://doc.qt.io/qt-5/qabstractitemmodel.html#data and it's notion of data roles, which has the same distinction between editable and text.

To make things concrete.  If you have an underlying float value, then you would:

- the editable value would be the actual float value, and you would be getting and setting Python floats.  It is the value given to the embedded cell editor (except on wxPython, because we haven't written a custom renderer yet).
- the text value is a str representing the float (ie. "12.05", etc.) possibly with formatting to a certain precision, or localization, etc.  It is what is displayed when the value is _not_ being edited.  I'd like to get rid of `set_text` in the future, but we need it for wx support right now.  So the current reason for `set_text` is "for compatibility with toolkits which have more limited editing capabilities".

As for why we are passing the model and indices, it is because it is strictly more powerful than the alternative.  The original intent was to allow things like alternate row shading (eg. `get_color` returns different values depending on the value of `row[-1] % 2`), but currently the plan is to push those up a layer to the to-be-developed view API, which _will_ be working in this way.

As an example of the sort of thing which the current API permits, but would be impossible with a simple to-from API, consider something like:
```
def get_text(self, model, row, column):
    if self.model.can_have_children(row, column):
        return "{} ({} children)".format(
            self.model.get_value(row, column),
            self.model.row_count(row, column)
        )
    else:
        return str(self.model.get_value(row, column))
```
I don't want to rule this sort of use-case out at this point, particularly given that you can write a class which wraps a to/from API in the more general API.

Reply:
Am I understanding correctly that this `AbstractValueType` is going to have more methods e.g. `get_color`, `set_color`, `get_image`, `set_image`?

The code that uses an interface should be closed for modifications. This `AbstractValueType` is intended to be an interface. One should be able to extend functionality by providing another implementation of the interface.  If we need to add more methods to support different data types, then the code that uses the interface will also need to change and call different methods depending on the data type. That does not sound right.

Maybe I am still misunderstanding this because I have not finished reading the entire thing. I will carry on and come back to this point. I hope it is alright to post my possibly premature thought.

Reply:
> Maybe I am still misunderstanding this because I have not finished reading the entire thing. I will carry on and come back to this point. I hope it is alright to post my possibly premature thought.

This is referring to the question about `get_text`/`set_text`/`get_editable`/`set_editable`/... (i.e. still unresolved for me... I need to do more work to understand).

The other half of my original comment regarding why model and indices are passed is resolved. Thank you.

Reply:
Yes, you are correct, it will grow other methods.  Interfaces are allowed to evolve during development, since no external code should be depending on them at this point.  Once the code is released I don't expect that the `AbstractValueType` will change from its final form.